pkg/chain/chain.go

```diff
@@ -1,164 +1,239 @@
-// Package eth contains interface for interaction with an ethereum blockchain
-// along with structures reflecting events emitted on an ethereum blockchain.
-package eth // TODO: rename; this can be any host chain
+// Package chain contains interface for interaction with a host chain along with
+// structures reflecting events emitted on such a chain.
+package chain
 
 import (
+	cecdsa "crypto/ecdsa"
+	"fmt"
 	"math/big"
 	"time"
 
-	"github.com/ethereum/go-ethereum/common"
 	"github.com/keep-network/keep-common/pkg/subscription"
 	"github.com/keep-network/keep-core/pkg/chain"
 	"github.com/keep-network/keep-ecdsa/pkg/ecdsa"
 )
 
-// Handle represents a handle to an ethereum blockchain.
+type KeepApplicationID interface {
+	fmt.Stringer
+}
+
+type KeepMemberID interface {
+	fmt.Stringer
+
+	// KeepMemberIDs should be convertible to their associated OperatorID. This
+	// may be a 1-to-1 correspondence, i.e. a simple interface cast, or it may
+	// be an internally-tracked association on the host chain.
+	OperatorID() OperatorID
+}
+
+type OperatorID interface {
+	fmt.Stringer
+
+	// OperatorIDs should be convertible to an associated KeepMemberID given the
+	// id of the keep in question. This may be a 1-to-1 correspondence, i.e. a
+	// simple interface cast, or it may be an internally-tracked association on
+	// the host chain. The conversion _may but is not guaranteed to be_ nil if
+	// the operator is not in the specified keep.
+	KeepMemberID(keepID KeepID) KeepMemberID
+}
+
+// Should be chain-specific, unique across chains, and simple ([A-Za-z0-9-_.]).
+// This can generally mean prefixing the common name of the chain followed by a
+// dash followed by the chain-specific identifier (e.g., the contract address
+// for Ethereum). For legacy reasons, Ethereum KeepIDs should always be the bare
+// Keep contract address.
+//
+// Should not be expected to be comparable between types or amongst themselves.
+type KeepID interface {
+	fmt.Stringer
+}
+
+// package chain in keep-core -- or keep-common/chain?
+// type BaseHandle interface {
+// 	BlockCounter() (BlockCounter, error)
+// 	StakeMonitor() (StakeMonitor, error)
+// 	BalanceMonitor() (BalanceMonitor, error)
+// }
+
+// Handle represents a handle to an implementation of operator functions
+// for a particular chain.
 type Handle interface {
-	// Address returns client's ethereum address.
-	Address() common.Address // TODO: use implementation-agnostic type
+	// Name returns the name of this chain handle. The name should be
+	// descriptive of the chain this handle represents; for example, for
+	// Ethereum it might return "Ethereum", for Celo "Celo". It can optionally
+	// include additional information if it would be useful in logs or debug
+	// output; for example, Ethereum handles might include network id
+	// information.
+	Name() string
```

Comment:
Would it be worth adding a `ChainType() string` if we decide to append additional information to the chain name in `Name()` function?

Reply:
I couldn't think of a good use case, but it's def a possibility.

I think if we went `ChainType` we might want to make it an enum/`iota`-based constant set, though.

Reply:
> I couldn't think of a good use case, but it's def a possibility.

It depends if we will end up appending additional information to the `Name`. If yes, then in most cases it won't be necessary to log the whole Name but just a chain type (I love the idea with enum).

Reply:
Worth flagging, I'm always thinking of log messages as I receive them as someone who isn't running a given node. For example, if every log message included the chain id, I wouldn't have to ask if a given client is running on mainnet or Ropsten. This is part of why I don't want to make the interface overly broad here---I think including the full `Name` is potentially a win, and it's up to the chain to decide what the relevant details are for that name.

I agree we could end up overbloating the name field though (I'm looking at you, web browser user agents). It feels like the right move might be starting with `Name`, and adding `ChainType` if we find it's necessary.

Reply:
Oh, now I see what you meant in the code comment. It makes perfect sense!